ons-spark/ancillary-topics/pandas-udfs.ipynb

```diff
@@ -0,0 +1,931 @@
+       "7          255.0  \n",
+       "8          255.0  \n",
+       "9          255.0  "
+      ]
+     },
+     "execution_count": 27,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import pandas as pd\n",
+    "from pyspark.sql import SparkSession\n",
+    "import pyspark.sql.functions as F\n",
+    "from pyspark.sql.functions import pandas_udf\n",
+    "import pyspark.sql.types as T\n",
+    "\n",
+    "spark = (SparkSession.builder.master(\"local[2]\")\n",
+    "         .appName(\"pandas_udfs\")\n",
+    "         .config('spark.sql.execution.arrow.enabled', 'true')\n",
+    "         .config('spark.sql.execution.arrow.fallback.enabled', 'true')\n",
+    "         .config('spark.excutorEnv.ARROW_PRE_0_15_IPC_FORMAT', 1)\n",
+    "         .config('spark.workerEnv.ARROW_PRE_0_15_IPC_FORMAT', 1)\n",
+    "         .getOrCreate())\n",
+    "\n",
+    "import yaml\n",
+    "with open(\"../../config.yaml\") as f:\n",
+    "     config = yaml.safe_load(f)\n",
+    "\n",
+    "rescue_path_csv = config[\"rescue_path_csv\"]\n",
+    "rescue = spark.read.csv(rescue_path_csv, header=True, inferSchema=True)\n",
+    "\n",
+    "rescue = rescue.withColumn(\n",
+    "    \"IncidentDuration\", \n",
+    "    F.col(\"PumpHoursTotal\") / F.col(\"PumpCount\")\n",
+    ")\n",
+    "\n",
+    "rescue = rescue.select(F.col('IncidentNumber').alias('incident_number'),\n",
+    "                       F.col('AnimalGroupParent').alias('animal_group'), \n",
+    "                       F.col('IncidentDuration').alias('incident_duration'), \n",
+    "                       F.col('IncidentNotionalCost(£)').alias('incident_cost'))\n",
+    "\n",
+    "rescue = rescue.filter(F.col('incident_duration').isNotNull())\n",
+    "\n",
+    "rescue.limit(10).toPandas()"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Scalar Pandas UDFs\n",
+    "\n",
+    "Scalar Pandas UDFs take in a `pandas.Series` and output a `pandas.Series`. \n",
+    "\n",
+    "Spark executes scalar Pandas UDFs by serialising each partition column into a `pandas.Series` object (basically splitting the data into batches). The UDF is then called on each of of these Series objects as a subset of the data and then results are concactenated together and returned as a `pandas.Series`. \n",
+    "\n",
+    "Be aware that scalar UDFs may cause incorrect results (if your dataframe is over multiple partitions) when calculating means and standard deviations. For more info on this and how to overcome this issue see [this towards data science post](https://towardsdatascience.com/pyspark-or-pandas-why-not-both-95523946ec7c).\n",
+    "\n",
+    "In the decorator of a scalar Pandas UDF the first arguement is the data type of the output dataframe and the second argument is the UDF type. Here is a simple Pandas scalar UDF to find out how much each minute cost per incident."
```

Comment:
this calculation doesn't make a whole load of sense because there is a already a column for `HourlyNotionalCost`, but that's not important. We just need the an example of the scalar udf